src/main/java/org/opensearch/neuralsearch/highlight/NeuralHighlighterEngine.java

```diff
@@ -0,0 +1,306 @@
+     */
+    public String extractOriginalQuery(Query query, String fieldName) {
+        if (fieldName == null) {
+            log.warn("Field name is null, extraction may be less accurate");
+        }
+        return queryTextExtractorRegistry.extractQueryText(query, fieldName);
+    }
+
+    /**
+     * Gets the model ID from the options
+     *
+     * @param options The options map
+     * @return The model ID
+     */
+    public String getModelId(Map<String, Object> options) {
+        Object modelId = options.get(MODEL_ID_FIELD);
+        if (Objects.isNull(modelId)) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "Missing required option: %s", MODEL_ID_FIELD));
+        }
+        return modelId.toString();
+    }
+
+    /**
+     * Gets highlighted text from the ML model
+     *
+     * @param modelId The ID of the model to use
+     * @param question The search query
+     * @param context The document text
+     * @return Formatted text with highlighting
+     */
+    public String getHighlightedSentences(String modelId, String question, String context) {
+        List<Map<String, Object>> result = fetchModelResults(modelId, question, context);
+        if (result == null) {
+            return StringUtils.EMPTY;
+        }
+
+        return applyHighlighting(context, result);
+    }
+
+    /**
+     * Fetches highlighting results from the ML model
+     *
+     * @param modelId The ID of the model to use
+     * @param question The search query
+     * @param context The document text
+     * @return The highlighting results
+     */
+    public List<Map<String, Object>> fetchModelResults(String modelId, String question, String context) {
+        PlainActionFuture<List<Map<String, Object>>> future = PlainActionFuture.newFuture();
+
+        SentenceHighlightingRequest request = SentenceHighlightingRequest.builder()
+            .modelId(modelId)
+            .question(question)
+            .context(context)
+            .build();
+
+        mlCommonsClient.inferenceSentenceHighlighting(request, future);
+
+        try {
+            return future.actionGet();
+        } catch (Exception e) {
+            throw new OpenSearchException(
+                String.format(Locale.ROOT, "Error during sentence highlighting inference from model [%s]", modelId),
+                e
+            );
+        }
+    }
+
+    /**
+     * Applies highlighting to the original context based on the ML model response
+     *
+     * @param context The original document text
+     * @param highlightResults The highlighting results from the ML model
+     * @return Formatted text with highlighting
+     */
+    public String applyHighlighting(String context, List<Map<String, Object>> highlightResults) {
+        List<int[]> validHighlights = new ArrayList<>();
+
+        if (highlightResults != null && !highlightResults.isEmpty()) {
+            Map<String, Object> result = highlightResults.getFirst();
```

Comment:
If we are only handling the very first item from the list, the method should receive single item instead of list.
```
public String applyHighlighting(String context, Map<String, Object> highlightResults) {
```